src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -491,7 +499,38 @@ static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
   const dereference_exprt obj_deref(pointer2, class_type);
 
   return member_exprt(
-    obj_deref, fieldref.get(ID_component_name), fieldref.type());
+    obj_deref,
+    fieldref.get(ID_component_name),
+    fieldref.type());
+}
+
+codet java_bytecode_convert_methodt::get_array_bounds_check(
+  const exprt &arraystruct,
+  const exprt &idx,
+  const source_locationt& original_sloc)
+{
+  constant_exprt intzero=as_number(0, java_int_type());
+  binary_relation_exprt gezero(idx, ID_ge, intzero);
+  const member_exprt length_field(arraystruct, "length", java_int_type());
+  binary_relation_exprt ltlength(idx, ID_lt, length_field);
+  code_blockt bounds_checks;
+
+  bounds_checks.add(code_assertt(gezero));
+  bounds_checks.operands().back().add_source_location()=original_sloc;
+  bounds_checks.operands().back().add_source_location()
+    .set_comment("Array index < 0");
+  bounds_checks.operands().back().add_source_location()
+    .set_property_class("array-index-oob-low");
+  bounds_checks.add(code_assertt(ltlength));
+
+  bounds_checks.operands().back().add_source_location()=original_sloc;
+  bounds_checks.operands().back().add_source_location()
+    .set_comment("Array index >= length");
+  bounds_checks.operands().back().add_source_location()
+    .set_property_class("array-index-oob-high");
+
+  // TODO make this throw ArrayIndexOutOfBoundsException instead of asserting.
```

Comment:
maybe I am wrong, but doesn't the code actually generate both (0<=index< length) ?
```
  bounds_checks.add(code_assertt(gezero));
  bounds_checks.operands().back().add_source_location()=original_sloc;
  bounds_checks.operands().back().add_source_location()
    .set_comment("Array index < 0");
  bounds_checks.operands().back().add_source_location()
    .set_property_class("array-index-oob-low");
  bounds_checks.add(code_assertt(ltlength));

  bounds_checks.operands().back().add_source_location()=original_sloc;
  bounds_checks.operands().back().add_source_location()
    .set_comment("Array index >= length");
  bounds_checks.operands().back().add_source_location()
    .set_property_class("array-index-oob-high");
```